File: Dynamic Programming/Jump Game V/solution_O(n).py
class Solution:
    def maxJumps(self, arr: List[int], d: int) -> int:
        n = len(arr)
        next_higher = [n for i in range(n)]
        prev_higher = [n for i in range(n)]
        
        stack = []
        for i in range(n):
            while len(stack) > 0 and i - stack[-1] <= d and arr[i] > arr[stack[-1]]:
                next_higher[stack.pop()] = i
            stack.append(i)
        
        for i in range(n-1, -1, -1):
            while len(stack) > 0 and stack[-1] - i <= d and arr[i] > arr[stack[-1]]:
                prev_higher[stack.pop()] = i
            stack.append(i)
            
        @cache
        def dfs(idx):
            if idx >= n:
                return -1
            return max(1 + dfs(next_higher[idx]), 1 + dfs(prev_higher[idx]))
        result = 0
        for i in range(n):
            result = max(result, 1 + dfs(i))
        
        return result

File: Dynamic Programming/Jump Game V/solution_O(n).cpp
class Solution {
public:
    int maxJumps(vector<int>& arr, int d) {
        int n = arr.size();
        vector<int> next_higher(n, n), prev_higher(n, n);
        vector<int> st;
        for(int i = 0; i < n; ++i) {
            while(!st.empty() && i - st.back() <= d && arr[i] > arr[st.back()]) {
                int j = st.back(); st.pop_back();
                next_higher[j] = i;
            }
            st.push_back(i);
        }
        st.clear();
        for(int i = n-1; i >= 0; --i) {
            while(!st.empty() && st.back() - i <= d && arr[i] > arr[st.back()]) {
                int j = st.back(); st.pop_back();
                prev_higher[j] = i;
            }
            st.push_back(i);
        }
        vector<int> dp(n, -1);
        auto dfs = [&](this auto& dfs, int u) {
            if(u == n) return 0;
            if(dp[u] != -1) return dp[u];
            return dp[u] = 1 + max(dfs(next_higher[u]), dfs(prev_higher[u]));
        };
        for(int i = 0; i < n; ++i) dfs(i);
        return *max_element(begin(dp), end(dp));
    }
};
